lib/pbench/server/database/models/users.py

```diff
@@ -1,58 +1,187 @@
+    key; the cause will specify the original SQLAlchemy exception.
+    """
+
+    def __init__(self, operation: str, params: JSONOBJECT, cause: str):
+        self.operation = operation
+        self.params = params
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Error {self.operation} {self.params!r}: {self.cause}"
+
+
+class UserDuplicate(UserError):
+    """Attempt to commit a duplicate unique value."""
+
+    def __init__(self, user: "User", cause: str):
+        self.user = user
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Duplicate user setting in {self.user.as_json()}: {self.cause}"
+
+
+class UserNullKey(UserError):
+    """Attempt to commit an Audit row with an empty required column."""
+
+    def __init__(self, user: "User", cause: str):
+        self.user = user
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Missing required key in {self.user.as_json()}: {self.cause}"
+
+
+class UserProfileBadKey(UserError):
+    """An unsupported profile key was specified.
+
+    The error text will identify the profile key that was specified.
+    """
+
+    def __init__(self, key: str):
+        self.key = key
+
+    def __str__(self) -> str:
+        return f"User profile key {self.key!r} is not supported"
+
+
+class UserProfileBadStructure(UserError):
+    """Invalid Profile structure encountered.
+
+    A call to update found a level in the JSON document where
+    the caller's key expected a nested JSON object but the type at that level
+    is something else. For example, when `profile.user.email` finds that
+    `profile.user` is a string, it's impossible to look up the `email` field.
+
+    The error text will identify the key path and the expected key element that
+    is missing.
+    """
+
+    def __init__(self, key: str):
+        self.key = key
+
+    def __str__(self) -> str:
+        return f"Key value for {self.key!r} in profile is not a JSON object"
+
+
 class User(Database.Base):
     """User Model for storing user related details."""
 
     __tablename__ = "users"
 
+    # User profile keys
+    #
+    # SERVER is an internally maintained key namespace for additional metadata
+    # relating to the server's management of user. The information here is
+    # accessible to callers, but can't be changed.
+    #
+    # {"server.role": ["ADMIN"]}
+    #
+    # USER provides an open namespace allowing a user which is authenticated
+    # to define arbitrary metadata accessible only to that authenticated user.
+    # Each user can have its own unique value for this key, for example
+    # "user.email"/"user.first_name".
+    # Unauthenticated users can neither set nor read the user profile
+    # in the USER namespace.
+    # Note: USER namespace will be populated by the Server when the user access
+    # the Pbench Server resources for the very first time.
+    #
+    # {"user.email": user@email.com}
```

Comment:
I don't think making the user profile public by default would be a good idea, at least we should ask them `do you wish to make your profile accessible to other users on the Pbench-Server?` if the answer is yes we make it public and if the answer is no it's only private. It can be easily done by adding `server.access="public"/"private"` but not sure if its worth doing that now. But keeping this user profile information with us gives us the ability to make it public later if the user wishes to.

Reply:
I don't see any point in a non-authoritative local cache of private user information. Yeah, sure, we can have the UI show a local profile; but who really cares if it's private? I know my email, and if nobody else can see it, there's no point in making me worry about it. I can change it with the OIDC provider (if it allows that). Why would we want to cache it?

Another way of looking at this: we can't inadvertently be responsible for exposing PII we don't have. 😁

Reply:
The only "cache" that I'm contemplating is the username associated with a UUID.  If we want to expand the translation table beyond those two columns, then, yeah, we could hang other stuff on the UUID.  The obvious first choice are "settings" (like default retention period), which are only of interest to the owning user.  Anything like first/last name and email address are only useful in terms of sharing them with other people (as Dave says, the user already knows those things).  So, if we want to go all social media and allow users to find out about other users ("Hey, I found this interesting public dataset owned by `drb`, and I want to find out who that is so that I can contact him!"), then there's a good argument for supporting a public profile, which a user could (voluntarily!) populate (and update and clear).  But, that's very much a "future", and, while it would be good to avoid doing anything today which prevents us from going in that direction later, I don't think we should be putting _anything_ into building foundations for it now!